crypto/async/arch/async_posix.c

```diff
@@ -38,13 +78,23 @@ int async_fibre_makecontext(async_fibre *fibre)
     fibre->env_init = 0;
 #endif
     if (getcontext(&fibre->fibre) == 0) {
-        fibre->fibre.uc_stack.ss_sp = OPENSSL_malloc(STACKSIZE);
+        size_t num = STACKSIZE;
+
+        /*
+         *  Disallow customisation after the first
+         *  stack is allocated.
+         */
+        if (allow_customize)
+            allow_customize = 0;
```

Comment:
There is an inherent race condition here. I realise this just mirrors the existing race condition in `CRYPTO_malloc`.

Reply:
I've now protected this with a mutex, and done the same in a separate commit for `CRYPTO_set_mem_functions`.

Reply:
I'm not sure how to solve this without a large performance impact.

Reply:
Yeah - tricky. I'm not sure the best answer either. Possibly we could just document the new `ASYNC_set_mem_functions` as non-thread safe and must be called before other threads call any `ASYNC_*` functions. Probably also add a comment here as well explaining that.

Reply:
I realised we could likely get away with:

```
        if (allow_customize) {
            if (!RUN_ONCE(&async_mem_init, do_async_mem_init))
                return 0;
            if (!CRYPTO_THREAD_write_lock(async_mem_lock))
                return 0;
            allow_customize = 0;
            CRYPTO_THREAD_unlock(async_mem_lock);
        }
 ```
 
Worst case a few calls to `async_fibre_makecontext` get serialized.  There's not really a performance penalty after that...

If this is acceptable I can open another PR for `CRYPTO_set_mem_functions`.